SofaKernel/modules/SofaImplicitOdeSolver/src/SofaImplicitOdeSolver/StaticSolver.cpp

```diff
@@ -62,145 +76,356 @@ void StaticSolver::parse(sofa::core::objectmodel::BaseObjectDescription* arg)
+        }
+
+        // At this point, we completed one iteration, increment the counter.
+        // The rest is only for convergence tests and logging.
+        n_it++;
+
+        // The rest of the step is only necessary when doing more than one Newton iteration. Otherwise, we will
+        // waste computation time to reassemble the residual and compute the norms for a convergence that will
+        // never happen (we will always reach the maximum number of iterations, which is 1)
+        if (max_number_of_newton_iterations == 1) {
+            converged = true; // Not really, but we won't warn about divergence when it is always the case
+            diverged = false;
+            break;
+        }
+
+        // Part IV. Update the force vector.
+        {
+            ScopedAdvancedTimer _t_("UpdateForce");
 
-            // Compute addForce, in mapped: addForce + applyJT (vec)
-            force.clear();
             mop.computeForce(force);
             mop.projectResponse(force);
-            double f_cur_norm = sqrt(force.dot(force));
+        }
 
-            dx_norm = sqrt(dx.dot(dx));
+        // Part V. Compute the updated norms.
+        {
+            ScopedAdvancedTimer _t_("ComputeNorms");
 
+            // Residual norm
+            R_squared_norm = force.dot(force);
 
-            msg_info() << "Newton iteration #" << n_it << ": |f - K(x0 + dx)| = " << f_cur_norm << " |dx| = " << dx_norm;
-            sofa::helper::AdvancedTimer::valSet("residual", f_cur_norm);
-            sofa::helper::AdvancedTimer::valSet("correction", dx_norm);
-            sofa::helper::AdvancedTimer::stepEnd(stepname.c_str());
+            if (n_it == 1) {
+                R0_squared_norm = R_squared_norm;
+                R_previous_squared_norm = R0_squared_norm;
+            }
 
-            if (d_should_diverge_when_residual_is_growing.getValue() && f_cur_norm > f_norm && n_it>1) {
-                msg_info() << "[DIVERGED] residual's norm increased";
+            p_squared_residual_norms.emplace_back(R_squared_norm);
+
+            // Displacement norm
+            U.peq(dx);
+            dx_squared_norm = dx.dot(dx);
+            U_squared_norm= U.dot(U);
+
+            p_squared_increment_norms.emplace_back(dx_squared_norm);
+        }
+
+        // Part VI. Stop timers and print step information.
+        {
+            auto iteration_time = duration_cast<nanoseconds>(steady_clock::now() - t).count();
+
+            if (print_log) {
+                info << "Newton iteration #" << std::left << std::setw(5) << n_it
+                     << std::scientific
+                     << "  |R| = " << std::setw(12) << std::sqrt(R_squared_norm)
+                     << "  |R|/|R0| = " << std::setw(12) << std::sqrt(R_squared_norm / R0_squared_norm)
+                     << "  |du| = " << std::setw(12) << std::sqrt(dx_squared_norm)
+                     << "  |du| / |U| = " << std::setw(12) << std::sqrt(dx_squared_norm / U_squared_norm)
```

Comment:
Possible division by zero: `U_squared_norm` could be zero
You check if `U_squared_norm < epsilon*epsilon` in the next Part.